draft-ietf-httpbis-http2-secondary-certs.md

```diff
@@ -470,6 +486,10 @@ authentication request identifier, `Request-ID`. A peer that receives a
 error of type `PROTOCOL_ERROR`. Frames with identical request
 identifiers refer to the same `CERTIFICATE_REQUEST`.
 
+The `CERTIFICATE_NEEDED` frame MUST NOT be sent by servers prior to seeing the
+end of the client's request, and then MUST NOT be sent if the client included a
```

Comment:
We need to be very clear about what it means to say that the request is ended, because the client won't be able to send USE_CERTIFICATE on a half-closed stream.  (Do we have an issue for that?)

Reply:
Yes, #482.

Reply:
Aside from the point raised by @martinthomson, I think we should allow a server to send a `CERTIFICATE_NEEDED` frame before seeing the end of the request.

As stated in https://github.com/MikeBishop/http2-certs/pull/17#discussion_r151328990, CGI (and other APIs that have been designed based on CGI) require the properties of a certificate (e.g. DN) to be sent before the request body.

Unless we allow the `CERTIFICATE_NEEDED` to be sent while the request body is being transmitted, a server will be required to buffer entire request body when it needs to enforce the use of a client certificate.

Reply:
That's a good point.  The reason for that text is to ensure there's not a race, so the server doesn't send in ignorance of whether the client already included one.  (And therefore, doesn't get either side confused about whether the `USE_CERTIFICATE` is in response to the server's `CERTIFICATE_REQUIRED`.)